Add NSPopover to _WKWebExtensionAction.

xeenon opened the pull request:
#### 75b646bd8fd82fcc3457d7c99efb23206a3617ca
<pre>
Add NSPopover to _WKWebExtensionAction.
<a href="https://webkit.org/b/270126">https://webkit.org/b/270126</a>
<a href="https://rdar.apple.com/problem/123655122">rdar://problem/123655122</a>

Reviewed by Jeff Miller.

Also remove the ContentSizeDidChange and DidClose notifications, since those things
and handled internally now with the UIViewController and NSPopover paths.

Rename closePopupWebView to closePopup to also better align with the new approach.

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.mm:
(-[_WKWebExtensionAction popupPopover]):
(-[_WKWebExtensionAction closePopup]):
(-[_WKWebExtensionAction closePopupWebView]): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(-[_WKWebExtensionActionWebViewDelegate webViewWebContentProcessDidTerminate:]):
(-[_WKWebExtensionActionWebViewDelegate webViewDidClose:]):
(-[_WKWebExtensionActionViewController _viewControllerDismissalTransitionDidEnd:]):
(-[_WKWebExtensionActionPopover initWithWebExtensionAction:]):
(-[_WKWebExtensionActionPopover dealloc]):
(-[_WKWebExtensionActionPopover popoverWillClose:]):
(-[_WKWebExtensionActionPopover popoverDidClose:]):
(-[_WKWebExtensionActionPopover _otherPopoverWillShow:]):
(WebKit::WebExtensionAction::clearCustomizations):
(WebKit::WebExtensionAction::setPopupPath):
(WebKit::WebExtensionAction::popupPopover):
(WebKit::WebExtensionAction::setPopupPopoverAppearance):
(WebKit::WebExtensionAction::detectPopoverColorScheme):
(WebKit::WebExtensionAction::popupWebView):
(WebKit::WebExtensionAction::readyToPresentPopup):
(WebKit::WebExtensionAction::popupSizeDidChange):
(WebKit::WebExtensionAction::closePopup):
(WebKit::WebExtensionAction::popupDidClose): Deleted.
(WebKit::WebExtensionAction::closePopupWebView): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
(WebKit::WebExtensionAction::popupPopoverAppearance const):
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_identifier_name_in_declaration): Make an exception for _WK prefixed classes.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/275391@main">https://commits.webkit.org/275391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99cae319720446567af3d9d68c372783caa82d2a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41736 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20750 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44118 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44312 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37829 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44043 "Failed to checkout and rebase branch from PR 25155") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18081 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42310 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/17683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/35955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/41607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/36979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45701 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/37300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/18170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5582 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -532,7 +633,96 @@ - (void)_dismissPopup
+
+void WebExtensionAction::setPopupPopoverAppearance(Appearance appearance)
+{
+    m_popoverAppearance = appearance;
```
any reason on why we want to set the appearance before checking if there's a popup?

xeenon replied:
Yeah, should guard this too. The check function bails early without a popover too.

xeenon replied:
Actually it is still useful since the appearance will be used next time the popover loads.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -394,6 +398,93 @@ - (void)_dismissPopup
+
+- (void)dealloc
+{
+    [NSNotificationCenter.defaultCenter removeObserver:self];
```
This isn't necessary, Foundation will do this automatically on all the macOS versions we care about.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -532,7 +633,96 @@ - (void)_dismissPopup
+    if (!m_popupPopover || !m_popupPopover.get().shown)
+        return;
+
+    static NSString *checkColorSchemeScript = @"(function() {"
```
```suggestion
    static NSString * const checkColorSchemeScript = @"(function() {"
```

jeffmapple commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -532,7 +633,96 @@ - (void)_dismissPopup
+    "  return null;"
+    "})()";
+
+    [m_popupWebView evaluateJavaScript:checkColorSchemeScript completionHandler:^(id result, NSError *error) {
```
You could log any `error` here.

jeffmapple commented on Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
```diff
@@ -68,6 +73,10 @@ class WebExtensionAction : public API::ObjectImpl<API::Object::Type::WebExtensio
     enum class FallbackWhenEmpty { No, Yes };
 
+#if PLATFORM(MAC)
+    enum class Appearance : uint8_t { Unknown, Light, Dark, Both };
```
Maybe `Default` is a better name than `Unknown`?

xeenon committed:
Add NSPopover to _WKWebExtensionAction.
https://webkit.org/b/270126
rdar://problem/123655122

Reviewed by Jeff Miller.

Also remove the ContentSizeDidChange and DidClose notifications, since those things
and handled internally now with the UIViewController and NSPopover paths.

Rename closePopupWebView to closePopup to also better align with the new approach.

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.mm:
(-[_WKWebExtensionAction popupPopover]):
(-[_WKWebExtensionAction closePopup]):
(-[_WKWebExtensionAction closePopupWebView]): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(-[_WKWebExtensionActionWebViewDelegate webViewWebContentProcessDidTerminate:]):
(-[_WKWebExtensionActionWebViewDelegate webViewDidClose:]):
(-[_WKWebExtensionActionViewController _viewControllerDismissalTransitionDidEnd:]):
(-[_WKWebExtensionActionPopover initWithWebExtensionAction:]):
(-[_WKWebExtensionActionPopover dealloc]):
(-[_WKWebExtensionActionPopover popoverWillClose:]):
(-[_WKWebExtensionActionPopover popoverDidClose:]):
(-[_WKWebExtensionActionPopover _otherPopoverWillShow:]):
(WebKit::WebExtensionAction::clearCustomizations):
(WebKit::WebExtensionAction::setPopupPath):
(WebKit::WebExtensionAction::popupPopover):
(WebKit::WebExtensionAction::setPopupPopoverAppearance):
(WebKit::WebExtensionAction::detectPopoverColorScheme):
(WebKit::WebExtensionAction::popupWebView):
(WebKit::WebExtensionAction::readyToPresentPopup):
(WebKit::WebExtensionAction::popupSizeDidChange):
(WebKit::WebExtensionAction::closePopup):
(WebKit::WebExtensionAction::popupDidClose): Deleted.
(WebKit::WebExtensionAction::closePopupWebView): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
(WebKit::WebExtensionAction::popupPopoverAppearance const):
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_identifier_name_in_declaration): Make an exception for _WK prefixed classes.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAction.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/275391@main
